metecho/api/serializers.py

```diff
@@ -356,15 +357,46 @@ def update(self, instance, validated_data):
                 org_type=SCRATCH_ORG_TYPES.Dev
             )
             for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+                if new_user := self._valid_reassign(
+                    "dev", instance.assigned_dev, validated_data["assigned_dev"]
+                ):
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                else:
+                    org.queue_delete(originating_user_id=originating_user_id)
         if instance.assigned_qa != validated_data["assigned_qa"]:
             orgs = instance.scratchorg_set.active().filter(
                 org_type=SCRATCH_ORG_TYPES.QA
             )
             for org in orgs:
-                org.queue_delete(originating_user_id=originating_user_id)
+                if new_user := self._valid_reassign(
+                    "qa", instance.assigned_qa, validated_data["assigned_qa"]
+                ):
+                    org.queue_reassign(
+                        new_user=new_user, originating_user_id=originating_user_id
+                    )
+                else:
+                    org.queue_delete(originating_user_id=originating_user_id)
         return super().update(instance, validated_data)
 
+    def _valid_reassign(self, type_, old_assignee, new_assignee):
+        old_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": old_assignee}
+        )
+        new_user = self.get_matching_assigned_user(
+            type_, {f"assigned_{type_}": new_assignee}
+        )
+        if old_user and new_user and old_user.sf_username == new_user.sf_username:
+            return new_user
+        return None
+
+    def get_matching_assigned_user(self, type_, validated_data):
+        assigned = validated_data.get(f"assigned_{type_}", {})
```

Comment:
Given how this is used, it seems overly complex. Why couldn't this fn just accept `assigned` directly from `_valid_reassign` (either `old_assignee` or `new_assignee`)? That is, why create the `{f"assigned_{type_}": old_assignee}` dict in `_valid_reassign` just to unwrap it here in `get_matching_assigned_user`?

Reply:
This was taken from another PR where I had reason to write the logic this way; I wanted it to be compatible with that logic once both are merged.